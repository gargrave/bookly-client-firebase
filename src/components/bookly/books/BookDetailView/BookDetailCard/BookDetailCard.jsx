// @flow
import React, { Fragment } from 'react';
import { func, object } from 'prop-types';

import type { Book } from '../../../../../globals/flowtypes';

import AuthorLink from '../../../authors/AuthorLink';
import Button from '../../../../common/Button';
import ButtonRow from '../../../../common/ButtonRow';
import Card from '../../../../common/Card/Card';
import CardSpacer from '../../../../common/Card/CardSpacer/CardSpacer';
import CardTextList from '../../../../common/Card/CardTextList/CardTextList';

import {
  bookBasicDates,
  bookOptionalDates,
  hasOptionalDates,
} from './helpers';

type Props = {
  book: Book,
  onBackClick: Function,
  onDeleteClick: Function,
  onEditClick: Function,
};

const BookDetailCard = ({
  book,
  onBackClick,
  onDeleteClick,
  onEditClick,
}: Props) => {
  const {
    title,
  } = book;

  return (
    <Card
      classes={['detail-card', 'book-detail-card']}
      header={title}
      hoverable={false}
    >
      <AuthorLink author={book.author} />

      {hasOptionalDates(book) &&
        <Fragment>
          <CardTextList textList={bookOptionalDates(book)} />
          <CardSpacer />
        </Fragment>
      }
      <CardTextList textList={bookBasicDates(book)} />
      <CardSpacer size='large' />

      <ButtonRow>
        <Button
          onClick={onEditClick}
          position="left"
          text="Edit"
          type="info"
        />
        <Button
          onClick={onDeleteClick}
          text="Delete"
          type="danger"
        />
        <Button
          onClick={onBackClick}
          position="right"
          text="Back"
          type="light"
        />
      </ButtonRow>
    </Card>
  );
};

BookDetailCard.propTypes = {
  book: object.isRequired,
  onBackClick: func.isRequired,
  onDeleteClick: func.isRequired,
  onEditClick: func.isRequired,
};

export default BookDetailCard;
