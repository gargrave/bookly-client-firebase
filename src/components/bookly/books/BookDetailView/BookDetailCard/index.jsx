// @flow
import React from 'react';
import { func, instanceOf, oneOfType, shape, string } from 'prop-types';
import { format } from 'date-fns';

import type { Book } from '../../../../../constants/flowtypes';

import AuthorLink from '../../../authors/AuthorLink';
import Button from '../../../../common/Button';
import ButtonRow from '../../../../common/ButtonRow';
import Card from '../../../../common/Card';
import CardSpacer from '../../../../common/Card/CardSpacer';
import CardTextList from '../../../../common/Card/CardTextList';

type Props = {
  book: Book,
  onBackClick: Function,
  onDeleteClick: Function,
  onEditClick: Function,
};

const bookDatesTextList = (book: Book) => {
  const {
    created,
    updated,
  } = book;

  return [
    { title: 'Added', value: format(created, 'MMM. DD, YYYY, HH:mm:ss') },
    { title: 'Updated', value: format(updated, 'MMM. DD, YYYY, HH:mm:ss') },
  ];
};

const BookDetailCard = ({
  book,
  onBackClick,
  onDeleteClick,
  onEditClick,
}: Props) => {
  const {
    title,
  } = book;

  return (
    <Card
      classes={['detail-card', 'book-detail-card']}
      header={title}
      hoverable={false}
    >
      <AuthorLink
        author={book.author}
      />

      <CardTextList textList={bookDatesTextList(book)} />
      <CardSpacer size='large' />

      <ButtonRow>
        <Button
          onClick={onEditClick}
          position="left"
          text="Edit"
          type="info"
        />
        <Button
          onClick={onDeleteClick}
          text="Delete"
          type="danger"
        />
        <Button
          onClick={onBackClick}
          position="right"
          text="Back"
          type="light"
        />
      </ButtonRow>
    </Card>
  );
};

BookDetailCard.propTypes = {
  book: shape({
    created: oneOfType([
      instanceOf(Date),
      string,
    ]),
    updated: oneOfType([
      instanceOf(Date),
      string,
    ]),
    author: shape({
      firstName: string,
      lastName: string,
    }),
    title: string,
  }).isRequired,
  onBackClick: func.isRequired,
  onDeleteClick: func.isRequired,
  onEditClick: func.isRequired,
};

export default BookDetailCard;
