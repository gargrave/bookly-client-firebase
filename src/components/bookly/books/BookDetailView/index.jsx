// @flow
import React from 'react';
import { func, instanceOf, oneOfType, shape, string } from 'prop-types';
import { format } from 'date-fns';

import type { Book } from '../../../../constants/flowtypes';

import { buildClasses } from '../../../../utils/cssHelpers';

import Alert from '../../../common/Alert';
import AuthorLink from '../../authors/AuthorLink';
import Button from '../../../common/Button';
import ButtonRow from '../../../common/ButtonRow';
import Card from '../../../common/Card';

type Props = {
  book: Book,
  onBackClick: Function,
  onDeleteClick: Function,
  onEditClick: Function,
  topLevelError?: string,
};

function BookDetailView({
  book,
  onBackClick,
  onDeleteClick,
  onEditClick,
  topLevelError,
}: Props) {
  const {
    created,
    title,
    updated,
  } = book;

  return (
    <div className={buildClasses('book-detail-view')}>
      {topLevelError &&
        <Alert
          message={topLevelError}
          type="danger"
        />
      }

      <Card
        classes={['card--top-margin-med', 'detail-card', 'book-detail-card']}
        hoverable={false}
        title={title}
      >
        <AuthorLink
          author={book.author}
        />

        <hr/>
        <p className={buildClasses('card-text')}>
          <strong>Added:</strong> {format(created, 'MMM. DD, YYYY, HH:mm:ss')}
        </p>
        <p className={buildClasses('card-text')}>
          <strong>Updated:</strong> {format(updated, 'MMM. DD, YYYY, HH:mm:ss')}
        </p>

        <hr/>
        <ButtonRow>
          <Button
            onClick={onEditClick}
            position="left"
            text="Edit"
            type="info"
          />
          <Button
            onClick={onDeleteClick}
            text="Delete"
            type="danger"
          />
          <Button
            onClick={onBackClick}
            position="right"
            text="Back"
            type="light"
          />
        </ButtonRow>
      </Card>
    </div>
  );
}

BookDetailView.propTypes = {
  book: shape({
    created: oneOfType([
      instanceOf(Date),
      string,
    ]),
    updated: oneOfType([
      instanceOf(Date),
      string,
    ]),
    author: shape({
      firstName: string,
      lastName: string,
    }),
    title: string,
  }).isRequired,
  onBackClick: func.isRequired,
  onDeleteClick: func.isRequired,
  onEditClick: func.isRequired,
  topLevelError: string,
};

export default BookDetailView;
