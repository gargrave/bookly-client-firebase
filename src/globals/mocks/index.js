import { authorMocks } from './data/authorMocks';
import { bookBucketMocks } from './data/bookBucketMocks';
import { bookMocks } from './data/bookMocks';
import { profileMocks } from './data/profileMocks';
import { userMocks } from './data/userMocks';

export {
  authorMocks,
  bookBucketMocks,
  bookMocks,
  profileMocks,
  userMocks,
};
