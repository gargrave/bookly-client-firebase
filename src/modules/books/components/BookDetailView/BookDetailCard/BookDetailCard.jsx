// @flow
import React from 'react';
import { func, instanceOf, oneOfType, shape, string } from 'prop-types';
import { format } from 'date-fns';

import type { Book } from '../../../flowtypes';

import AuthorLink from '../../../../authors/components/AuthorLink/AuthorLink';
import Button from '../../../../common/components/Button/Button';
import ButtonRow from '../../../../common/components/ButtonRow/ButtonRow';
import Card from '../../../../common/components/Card/Card';

type Props = {
  book: Book,
  onBackClick: Function,
  onDeleteClick: Function,
  onEditClick: Function,
};

const bookDatesTextList = (book: Book) => {
  const {
    created,
    updated,
  } = book;

  return [
    { title: 'Added', value: format(created, 'MMM. DD, YYYY, HH:mm:ss') },
    { title: 'Updated', value: format(updated, 'MMM. DD, YYYY, HH:mm:ss') },
  ];
};

const BookDetailCard = ({
  book,
  onBackClick,
  onDeleteClick,
  onEditClick,
}: Props) => {
  const {
    title,
  } = book;

  return (
    <Card
      header={title}>
      <AuthorLink
        author={book.author} />

      <Card.TextList textList={bookDatesTextList(book)} />
      <Card.Spacer size='large' />

      <ButtonRow>
        <Button
          onClick={onEditClick}
          position="left"
          text="Edit"
          type="info" />
        <Button
          onClick={onDeleteClick}
          text="Delete"
          type="danger" />
        <Button
          onClick={onBackClick}
          position="right"
          text="Back"
          type="light" />
      </ButtonRow>
    </Card>
  );
};

BookDetailCard.propTypes = {
  book: shape({
    created: oneOfType([
      instanceOf(Date),
      string,
    ]),
    updated: oneOfType([
      instanceOf(Date),
      string,
    ]),
    author: shape({
      firstName: string,
      lastName: string,
    }),
    title: string,
  }).isRequired,
  onBackClick: func.isRequired,
  onDeleteClick: func.isRequired,
  onEditClick: func.isRequired,
};

export default BookDetailCard;
