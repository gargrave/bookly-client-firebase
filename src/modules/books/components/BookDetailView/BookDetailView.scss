@import '../../../../styles/_views';

.bookDetailView {
  @include viewWrapper;
}
