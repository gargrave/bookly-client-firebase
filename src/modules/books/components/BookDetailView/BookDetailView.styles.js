import styled from 'react-emotion'

import { views } from '../../../../styles'

export const BookDetailView = styled('div')`
  ${views.viewWrapper};
`
