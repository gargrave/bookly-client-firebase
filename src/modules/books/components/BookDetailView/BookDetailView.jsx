// @flow
import React from 'react'
import { func, instanceOf, oneOfType, shape, string } from 'prop-types'

import type { Book } from '../../flowtypes'

import Alert from '../../../common/components/Alert/Alert'
import BookDetailCard from './BookDetailCard/BookDetailCard'

import * as S from './BookDetailView.styles'

type Props = {
  book: Book,
  onBackClick: Function,
  onDeleteClick: Function,
  onEditClick: Function,
  topLevelError?: string,
}

const BookDetailView = ({
  book,
  onBackClick,
  onDeleteClick,
  onEditClick,
  topLevelError,
}: Props) => {
  return (
    <S.BookDetailView>
      {topLevelError && <Alert message={topLevelError} type="danger" />}
      <BookDetailCard
        book={book}
        onBackClick={onBackClick}
        onDeleteClick={onDeleteClick}
        onEditClick={onEditClick}
      />
    </S.BookDetailView>
  )
}

BookDetailView.propTypes = {
  book: shape({
    created: oneOfType([instanceOf(Date), string]),
    updated: oneOfType([instanceOf(Date), string]),
    author: shape({
      firstName: string,
      lastName: string,
    }),
    title: string,
  }).isRequired,
  onBackClick: func.isRequired,
  onDeleteClick: func.isRequired,
  onEditClick: func.isRequired,
  topLevelError: string,
}

export default BookDetailView
