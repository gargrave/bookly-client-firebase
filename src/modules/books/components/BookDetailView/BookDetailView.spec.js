import React from 'react';
import { shallow } from 'enzyme';

import { bookMocks } from '../../../../globals/mocks';

import BookDetailCard from './BookDetailCard';
import BookDetailView from './BookDetailView';

const defaultProps = Object.freeze({
  book: bookMocks[0],
  onBackClick: jest.fn(),
  onDeleteClick: jest.fn(),
  onEditClick: jest.fn(),
});

function getComponent(extraProps = {}) {
  const props = Object.assign({}, defaultProps, extraProps);
  return shallow(<BookDetailView {...props} />);
}

describe('BookDetailView', () => {
  let component;

  test('matches the snapshot', () => {
    component = getComponent();
    expect(component).toMatchSnapshot();
  });

  test('renders correctly', () => {
    component = getComponent();
    expect(component.find('.bookly-book-detail-view')).toHaveLength(1);
    expect(component.find(BookDetailCard)).toHaveLength(1);
  });
});
