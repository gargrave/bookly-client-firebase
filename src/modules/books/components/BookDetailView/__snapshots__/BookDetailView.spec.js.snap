// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BookDetailView matches the snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <BookDetailView
    book={
      Object {
        "author": Object {
          "bookCount": 1,
          "created": 2017-07-01T01:16:43.881Z,
          "firstName": "Aldous",
          "id": "oiw223h3qndlksadffl3ksdf",
          "lastName": "Huxley",
          "updated": 2017-07-01T01:16:43.881Z,
        },
        "created": 2018-01-14T01:36:26.220Z,
        "id": "sv3hmbefpc95MRXQsTHr",
        "sortBy": "",
        "title": "Brave New World",
        "updated": 2018-01-14T01:36:26.220Z,
      }
    }
    onBackClick={[Function]}
    onDeleteClick={[Function]}
    onEditClick={[Function]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": Array [
        undefined,
        <BookDetailCard
          book={
            Object {
              "author": Object {
                "bookCount": 1,
                "created": 2017-07-01T01:16:43.881Z,
                "firstName": "Aldous",
                "id": "oiw223h3qndlksadffl3ksdf",
                "lastName": "Huxley",
                "updated": 2017-07-01T01:16:43.881Z,
              },
              "created": 2018-01-14T01:36:26.220Z,
              "id": "sv3hmbefpc95MRXQsTHr",
              "sortBy": "",
              "title": "Brave New World",
              "updated": 2018-01-14T01:36:26.220Z,
            }
          }
          onBackClick={[Function]}
          onDeleteClick={[Function]}
          onEditClick={[Function]}
        />,
      ],
    },
    "ref": null,
    "rendered": Array [
      undefined,
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "book": Object {
            "author": Object {
              "bookCount": 1,
              "created": 2017-07-01T01:16:43.881Z,
              "firstName": "Aldous",
              "id": "oiw223h3qndlksadffl3ksdf",
              "lastName": "Huxley",
              "updated": 2017-07-01T01:16:43.881Z,
            },
            "created": 2018-01-14T01:36:26.220Z,
            "id": "sv3hmbefpc95MRXQsTHr",
            "sortBy": "",
            "title": "Brave New World",
            "updated": 2018-01-14T01:36:26.220Z,
          },
          "onBackClick": [Function],
          "onDeleteClick": [Function],
          "onEditClick": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": Array [
          undefined,
          <BookDetailCard
            book={
              Object {
                "author": Object {
                  "bookCount": 1,
                  "created": 2017-07-01T01:16:43.881Z,
                  "firstName": "Aldous",
                  "id": "oiw223h3qndlksadffl3ksdf",
                  "lastName": "Huxley",
                  "updated": 2017-07-01T01:16:43.881Z,
                },
                "created": 2018-01-14T01:36:26.220Z,
                "id": "sv3hmbefpc95MRXQsTHr",
                "sortBy": "",
                "title": "Brave New World",
                "updated": 2018-01-14T01:36:26.220Z,
              }
            }
            onBackClick={[Function]}
            onDeleteClick={[Function]}
            onEditClick={[Function]}
          />,
        ],
      },
      "ref": null,
      "rendered": Array [
        undefined,
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "book": Object {
              "author": Object {
                "bookCount": 1,
                "created": 2017-07-01T01:16:43.881Z,
                "firstName": "Aldous",
                "id": "oiw223h3qndlksadffl3ksdf",
                "lastName": "Huxley",
                "updated": 2017-07-01T01:16:43.881Z,
              },
              "created": 2018-01-14T01:36:26.220Z,
              "id": "sv3hmbefpc95MRXQsTHr",
              "sortBy": "",
              "title": "Brave New World",
              "updated": 2018-01-14T01:36:26.220Z,
            },
            "onBackClick": [Function],
            "onDeleteClick": [Function],
            "onEditClick": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;
